internal/pkg/summary/mustgather.go

```diff
@@ -0,0 +1,319 @@
+// relative file path when extracting the file or mapping in the counters.
+// OPCT collects must-gather automatically saving in the directory must-gather-opct.
+func (mg *MustGather) extractRelativePath(file string) string {
+	re := regexp.MustCompile(`must-gather-opct/([A-Za-z0-9]+(-[A-Za-z0-9]+)+\/)`)
+
+	split := re.Split(file, -1)
+	if len(split) != 2 {
+		return file
+	}
+	return split[1]
+}
+
+// extract dispatch to process must-gather items.
+func (mg *MustGather) extract(tarball *tar.Reader) error {
+
+	// Create must-gather directory
+	if _, err := os.Stat(mg.path); err != nil {
+		if err := os.MkdirAll(mg.path, 0755); err != nil {
+			return err
+		}
+	}
+
+	// Leaky bucket implementation (queue limit) to parallel process must-gather items
+	// without exhausting resources.
+	// Benckmark info: this parallel processing decreased 3 times the total processing time.
+	// Samples: Serial=~100s, rate(100)=~30s, rate(150)=~25s.
+	waiterProcNS := &sync.WaitGroup{}
+	chProcNSErrors := make(chan *MustGatherLog, 30)
+	semaphore := make(chan struct{}, 50)
+	// have a max rate of N/sec
+	rate := make(chan struct{}, 100)
+	for i := 0; i < cap(rate); i++ {
+		rate <- struct{}{}
+	}
+	// leaky bucket
+	go func() {
+		ticker := time.NewTicker(100 * time.Millisecond)
+		defer ticker.Stop()
+		for range ticker.C {
+			_, ok := <-rate
+			// if this isn't going to run indefinitely, signal
+			// this to return by closing the rate channel.
+			if !ok {
+				return
+			}
+		}
+	}()
+	// consumer
+	go func() {
+		for mgLog := range chProcNSErrors {
+			err := mg.processNamespaceErrors(mgLog)
+			if err != nil {
+				log.Errorf("error extracting log info: %v", err)
+			}
+			waiterProcNS.Done()
+		}
+	}()
+
+	// Walk through files in must-gather tarball file.
+	for {
+		header, err := tarball.Next()
+
+		switch {
+		// no more files
+		case err == io.EOF:
+			return nil
+
+		// return on error
+		case err != nil:
+			return err
+
+		// skip it when the headr isn't set (not sure how this happens)
+		case header == nil:
+			continue
+		}
+
+		// the target location where the dir/file should be created.
+		target := filepath.Join(mg.path, header.Name)
+		ok, typ := mg.matchToExtract(target)
+		if !ok {
+			continue
+		}
+		targetAlias := mg.extractRelativePath(target)
+
+		// the following switch could also be done using fi.Mode(), not sure if there
+		// a benefit of using one vs. the other.
+		// fi := header.FileInfo()
+
+		// check the file type
+		switch header.Typeflag {
+
+		// directories in tarball
+		case tar.TypeDir:
+
+			// creating subdirectories structures will be ignored and need
+			// sub-directories under mg.path must be created previously if needed.
+			/*
+				targetDir := filepath.Join(mg.path, targetAlias)
+				if _, err := os.Stat(targetDir); err != nil {
+					if err := os.MkdirAll(targetDir, 0755); err != nil {
+						return err
+					}
+				}
+			*/
+			break
```

Comment:
Don't want to create directory for each item in the tarball, we'll choose what to filter, and where to save in the next case item.

Suggestion to move that case item to bottom, keeping the example of creating nested directories for directory items in the tarball, continuing the loop in the case - or just remove the case item. :)